### ISSUE TITLE
[shorten-filenames] Problems with filenames that are too long for some systems

### ISSUE THREAD

**Issue by recursionbane** [coursera-dl/coursera-dl#414]
**OS**: Windows 7 Enterprise (64-bit)
**Course Type**: On-demand

**Usage**:
```
coursera-dl -u <USERNAME> -p <PASSWORD> --on-demand corporate-finance [--skip-download]
```

**Error trace**:
```
Downloading class: corporate-finance
Starting new HTTPS connection (1): www.coursera.org
Logged in on coursera.org.
Downloaded https://www.coursera.org/api/opencourse.v1/course/corporate-finance (39365 bytes)
Parsing syllabus of on-demand course. This may take some time, be patient ...
corporate-finance\01_course-overview\01_about-the-course-and-your-classmates\01_welcome-to-financial-evaluation-and-strategy-corporate-finance.mp4 already downloaded
corporate-finance\01_course-overview\01_about-the-course-and-your-classmates\01_welcome-to-financial-evaluation-and-strategy-corporate-finance.srt already downloaded
corporate-finance\02_module-1-the-objective-of-the-corporation-and-analysis-of-financial-ratios\01_lesson-1-1-objectives-and-overview\01_objectives-and-overview.mp4 already downloaded
corporate-finance\02_module-1-the-objective-of-the-corporation-and-analysis-of-financial-ratios\01_lesson-1-1-objectives-and-overview\01_objectives-and-overview.srt already downloaded
Traceback (most recent call last):
  File "C:\programming\python3\Scripts\coursera-dl-script.py", line 9, in <module>
    load_entry_point('coursera==0.0.3', 'console_scripts', 'coursera-dl')()
  File "c:\programming\python3\lib\site-packages\coursera\coursera_dl.py", line 942, in main
    result = download_on_demand_class(args, class_name)
  File "c:\programming\python3\lib\site-packages\coursera\coursera_dl.py", line 918, in download_on_demand_class
    args.intact_fnames
  File "c:\programming\python3\lib\site-packages\coursera\coursera_dl.py", line 516, in download_lectures
    open(lecfn, 'w').close()  # touch
FileNotFoundError: [Errno 2] No such file or directory: 'corporate-finance\\02_module-1-the-objective-of-the-corporation-and-analysis-of-financial-ratios\\02_lesson-1-2-why-maximizing-the-stock-price-is-a-reasonable-goal-for-a-company\\01_why-maximizing-the-stock-price-is-a-reasonable-goal-for-a-company.mp4'
```

Is this related to long filenames being a problem on Windows?

**Comment by balta2ar** [comment]
Most probably it is!

``` python
In [1]: len('corporate-finance\\02_module-1-the-objective-of-the-corporation-and-analysis-of-financial-ratios\\02_lesson-1-2-why-maximizing-the-stock-price-is-a-reasonable-goal-for-a-company\\01_why-maximizing-the-stock-price-is-a-reasonable-goal-for-a-company.mp4')                                                                                 
Out[1]: 248
```

**Comment by recursionbane** [comment]
Is there an option to shorten/truncate folder and file names as they are created?

**Comment by balta2ar** [comment]
I'm afraid, no. I expected to find an option that allows to override path formatting pattern, but it looks like patterns are hardcoded at the moment. We need to address this problem.

@rbrito What are our options? Custom format argument? So that users could write something like this: `--filename-format "{unit}/{module:20}"` and limit the length of the filename. What do you think?

**Comment by achaleus** [comment]
Hello I am facing same type error... any suggestions

FileNotFoundError: [Errno 2] No such file or directory: 'finiteelementmethods-001\\01_Unit_1-_Linear_elliptic_partial_differential_equations_in_one_dimension._Elasticity_heat_conduction_and_mass_diffusion\\01_01.01._Introduction._Linear_elliptic_partial_differential_equations_-_I_14-46.srt'

**Comment by aeunson** [comment]
Yep, its a path length problem.

I opened up the python file and found the problem in the download_lectures methods

in the version I have its around line 486

I just added a substring to limit the class name and formatted file name. Experiment with the numbers (60, 100) and make them smaller/larger as needed. (hack).

This was the simplest change I could make to get what I wanted (the course downloaded).

Good luck.
Andy

#AE: Hack the substrings to limit the file/folder name lengths 
        sec = os.path.join(path, class_name[0:60],
                           format_section(secnum + 1, section, class_name, verbose_dirs)[0:100])

**Comment by achaleus** [comment]
Hi @aeunson can you explain in laymen term's, I tried the one you mentioned above but it is giving syntax error

**Comment by balta2ar** [comment]
@achaleus Please post here the error that your interpreter reports along with the exact changes you have made so that we can help you.

**Comment by rbrito** [comment]
Since this bug is so long-lived, I guess that I will accept any hack here, even if it is dirty, so that windows users can follow their courses.

Whatever works is fine with me, as long as it doesn't cripple Linux and other Unices.

**Comment by bIgBV** [comment]
Thing is, I just faced the same issue on OSX.

**Comment by rbrito** [comment]
@balta2ar, IMVHO, I think that we should just use a sanitization function/method/whatever like muzicbrainz's Picard does... In my experience, it has the best path shortening that I know of and it is written in Python, which means that we can probably steal it, as it is under a compatible license (if I'm not mistaken).

Giving users yet another knob to tune would be too much of a support headache... We can implement it as an advanced option and tell the users to not touch it, but I feel that the default should be sane and detect (at run time) if we are running under Windows, Unix, etc. and take appropriate measures.

Any comments? Am I missing something obvious?

**Comment by balta2ar** [comment]
> IMVHO, I think that we should just use a sanitization function/method/whatever like muzicbrainz's Picard does... In my experience, it has the best path shortening that I know of and it is written in Python, which means that we can probably steal it, as it is under a compatible license (if I'm not mistaken).

@rbrito Could you please elaborate how this program does the shortening? I haven't ever used it. Or maybe even point me to a specific class in the implementation that does it. Thanks!

**Comment by GlowingCrystallineEntity** [comment]
+1 -- having this issue as well, on Windows 7.

**Comment by GlowingCrystallineEntity** [comment]
For anyone who wants to hack this, I added the following (hack, which I would not suggest committing as any sort of solution) based on @aeunson 's comments above.  This is in the current github version of coursera-dl.py, in download_lectures(), just above the "if combined_section_lectures_nums:" that I've included in the snippet.

There seem to be two cases: Most of the time (for videos, it seems), "title" is null or empty; but sometimes (perhaps for pdfs and similar), it has a value -- and it's really the concatenation of "lecname" and "title" that has to be < 64 (on Windows), but an extra character (underscore) is added when the program concatenates them, so they need different truncating if "title" has a value.




                if len(title) > 0 and len(lecname) + len(title) > 58:
                    lecname = lecname[:58]
                    title = title[:58-len(lecname)]
                elif len(lecname) > 60:
                    lecname = lecname[:60]

                if combined_section_lectures_nums:
                    lecfn = os.path.join(
                        sec,
                        format_combine_number_resource(
                            secnum + 1, lecnum + 1, lecname, title, fmt))
                else:
                    lecfn = os.path.join(
                        sec, format_resource(lecnum + 1, lecname, title, fmt))

**Comment by GlowingCrystallineEntity** [comment]
(p.s. -- this might be related to what happened to me initially in #461 -- if a pdf or similar already has a name (in "title" it seems), maybe it would be better to just use that rather than pre-pending the lecture name as well.)

**Comment by GlowingCrystallineEntity** [comment]
(p.p.s. -- also, as a *total* hack, I also added 'lecfn = lecfn.replace("?", "_")' after the code snippet above, to deal with the issue of trying to create a filename with a question mark in it, such as I commented on #453.

This just gets you over the d/l failure, though.  These files should at least be turned into htmls, and/or the source URL saved so someone can manually d/l the resource later, or, best, if they can be identified as known sites with known (say) PDF downloads of their own, it would be nice to grab those.  That's probably an unwieldy amount of work, though.)

**Comment by GlowingCrystallineEntity** [comment]
(p.p.p.s -- a course with all these issues is "wharton-people-analytics" / "People Analytics" in their course catalog.)

**Comment by exowanderer** [comment]
After including the hacks from GlowingCrystalineEntity, I was still getting the same "file name too long" errors, but from a different location in the TraceBack.

For that reason, I added a (very dangerous) hack that might help some people above.  

In the file "downloaders.py", I add this snippet near LINE 370 (between for "progress.start()" and " f = open(filename, 'ab') if resume else open(filename, 'wb')"

"
            progress.start()

            ''' BEGINNING CUSTOM HACK  '''

            import os, statvfs
            maxFileNameLength = os.statvfs('/')[statvfs.F_NAMEMAX]

            lastChunk   = filename.split('/')[-1]
            if len(lastChunk) > maxFileNameLength:
                filename = filename.replace(lastChunk, lastChunk[:maxFileNameLength])
            ''' END CUSTOM HACK '''
            
            f = open(filename, 'ab') if resume else open(filename, 'wb')
``

The whole point is to identify the actual file being edited, assess the length of it's name. If it is larger than `maxFileNameLength` (defined individually by your OS), then just chop off the end and move on with your downloads.  On my end, I also added a bunch of print() statements to identify which file was a problem.  I plan to hunt down those files manually later.
